### ISSUE TITLE
Relevance scores create spurious ties in low-precision evaluation

### ISSUE THREAD

**Issue by KisuYang** [embeddings-benchmark/mteb#4932]
### Describe the bug

When embedding/reranking models run in **float16/bfloat16**, MTEB computes the final relevance score in that same low precision in a couple of paths. Reduced-mantissa formats coarsely bucket values in the narrow score range, so **distinct query–document scores collapse into identical values ("spurious ties")**. Ranking then falls back to arbitrary tie-breaking (index order), which inflates the variance and bias of retrieval metrics and can even flip model rankings.

The standard `SentenceTransformerEncoderWrapper` already upcasts embeddings to float32 on return, so most bi-encoder leaderboard runs are fine. Two gaps remain on current `main`:

1. **Similarity primitives** — `mteb/similarity_functions.py::_convert_to_tensor` converts *non-tensor* inputs to float32 but leaves existing tensors at their incoming dtype. An encoder that returns bf16/fp16 tensors is therefore scored in low precision by `cos_sim` / `dot_score` / `max_sim`.
2. **Cross-encoder rerankers** — `qwen3_reranker.py` and `rerankers_monot5_based.py` apply `log_softmax` directly to logits from a bf16/fp16 model. (`kalm_reranker.py` already calls `.float()` first, which is exactly the fix.)

This is analyzed in "Reliable Evaluation Protocol for Low-Precision Retrieval" (Yang et al., 2026), https://aclanthology.org/2026.acl-short.33. The remedy — **High-Precision Scoring (HPS)** — upcasts *only* the final scoring op to float32; it is value-preserving (every fp16/bf16 value is exactly representable in fp32), zero-parameter, and a no-op for already-float32 embeddings, so full-precision leaderboard numbers are unchanged.

### To reproduce

```python
import torch
qn = torch.nn.functional.normalize
q, d = torch.randn(1, 32), torch.randn(200, 32)

# bi-encoder cosine score computed in bf16 (no HPS) vs upcast-then-score (HPS)
naive = qn(q.bfloat16(), dim=1) @ qn(d.bfloat16(), dim=1).T
hps   = qn(q.bfloat16().float(), dim=1) @ qn(d.bfloat16().float(), dim=1).T
print(torch.unique(naive).numel(), torch.unique(hps).numel())  # e.g. 184 vs 200

# reranker softmax over 200 two-logit rows, bf16 vs HPS
logits = torch.randn(200, 2).bfloat16()
p_bf16 = torch.nn.functional.log_softmax(logits, dim=1)[:, 1].exp()
p_hps  = torch.nn.functional.log_softmax(logits.float(), dim=1)[:, 1].exp()
print(torch.unique(p_bf16).numel(), torch.unique(p_hps).numel())  # e.g. 145 vs 196
```
Fewer unique scores = spurious ties, which then get broken arbitrarily during ranking.

### Additional information

Proposed fix (HPS) at both sites: upcast fp16/bf16 embeddings to float32 in the similarity primitives, and upcast reranker logits to float32 before the softmax. I've prepared two small, separate PRs and am opening them alongside this issue. Happy to gate behind a flag instead if maintainers prefer.

Disclosure: I'm the first author of the referenced paper.

### Are you interested to contribute a fix for this bug?

Yes

**Comment by KisuYang** [comment]
Thanks, that's exactly what the two linked PRs do. #4933 covers the similarity primitives (`_convert_to_tensor`) and #4934 the reranker log-softmax path. Both are no-ops for fp32 runs. Reviews welcome!

**Comment by KisuYang** [comment]
Both parts of this are now merged: #4933 (similarity primitives, incl. the follow-up test guard in #5053) and #4934 (cross-encoder reranker logits). Closing. The same fixes are being proposed to sentence-transformers: [huggingface/sentence-transformers#3893](https://github.com/huggingface/sentence-transformers/issues/3894).
